includes/class-sensei-course-manual-enrolment-provider.php

```diff
@@ -0,0 +1,242 @@
+		Sensei_Course_Enrolment_Manager::trigger_course_enrolment_check( $user_id, $course_id );
+
+		return ! $this->is_enrolled( $user_id, $course_id );
+	}
+
+	/**
+	 * Adds enrolment for student but does not trigger recaculation.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 */
+	private function add_student_enrolment( $user_id, $course_id ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+		update_term_meta( $term->term_id, $this->get_enrolment_status_meta_key( $course_id ), time() );
+
+		Sensei_Utils::user_start_course( $user_id, $course_id );
+	}
+
+	/**
+	 * Manually withdraw a student in a course.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 */
+	private function remove_student_enrolment( $user_id, $course_id ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+		delete_term_meta( $term->term_id, $this->get_enrolment_status_meta_key( $course_id ) );
+	}
+
+	/**
+	 * Gets the meta key used to track manual enrolment status.
+	 *
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return string
+	 */
+	private function get_enrolment_status_meta_key( $course_id ) {
+		return self::META_PREFIX_MANUAL_STATUS . $course_id;
+	}
+
+	/**
+	 * Migrates legacy enrolment data.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 */
+	private function migrate_legacy_enrolment( $user_id, $course_id ) {
+		$course_progress_comment_id = Sensei_Utils::get_course_progress_comment_id( $course_id, $user_id );
+
+		$migration_log      = [];
+		$is_legacy_enrolled = ! empty( $course_progress_comment_id );
+
+		$migration_log['had_progress'] = $is_legacy_enrolled;
+
+		/**
+		 * Allows other providers to have an opinion about whether or not a user was enrolled pre-3.0.0.
+		 * This should only be called once per user/course just after upgrading from a pre-3.0.0 version of Sensei.
+		 *
+		 * @since 3.0.0
+		 *
+		 * @param bool $is_legacy_enrolled If the user was actually enrolled before 3.0.0 migration.
+		 * @param int  $user_id            User ID.
+		 * @param int  $course_id          Course post ID.
+		 */
+		$is_legacy_enrolled = apply_filters( 'sensei_is_legacy_enrolled', $is_legacy_enrolled, $user_id, $course_id );
+
+		$migration_log['is_enrolled'] = $is_legacy_enrolled;
+
+		if ( $is_legacy_enrolled ) {
+			$this->add_student_enrolment( $user_id, $course_id );
+		}
+
+		$this->set_migrated_legacy_enrolment_status( $user_id, $course_id, $migration_log );
+	}
+
+	/**
+	 * Check if a user's enrolment status needs to be migrated from pre-3.0.0 enrolment.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return bool
+	 */
+	private function needs_legacy_migration( $user_id, $course_id ) {
+		if ( ! get_option( 'sensei_enrolment_legacy' ) ) {
+			return false;
+		}
+
+		return ! $this->has_migrated_legacy_enrolment( $user_id, $course_id );
```

Comment:
We might also want to not migrate if the course progress comment date was after the `sensei_enrolment_legacy` time.